main/src/ca/uwaterloo/flix/language/ast/SyntaxTree.scala

```diff
@@ -0,0 +1,504 @@
+     */
+    case class Tree(tree: SyntaxTree.Tree) extends Child
+  }
+
+
+  sealed trait TreeKind {
+    def debug_name: Option[String] = None
+  }
+
+  /**
+   * Different kinds of syntax nodes in a [[SyntaxTree]].
+   * The only error kind that holds data is the special [[TreeKind.ErrorTree]].
+   */
+  object TreeKind {
+    /**
+     * A special error kind wrapping a [[Parser2Error]].
+     */
+    case class ErrorTree(error: Parser2Error) extends TreeKind
+
+    case object Annotations extends TreeKind
+
+    case object Argument extends TreeKind
+
+    case object ArgumentNamed extends TreeKind
+
+    case object Arguments extends TreeKind
+
+    case object Case extends TreeKind
+
+    case object Comments extends TreeKind
+
+    case object Doc extends TreeKind
+
+    case object Ident extends TreeKind
+
+    case object Kind extends TreeKind
+
+    case object Modifiers extends TreeKind
+
+    case object Operator extends TreeKind
+
+    case object Parameter extends TreeKind
+
+    case object Parameters extends TreeKind
+
+    case object QName extends TreeKind
+
+    case object Source extends TreeKind
+
+    case object TypeParameter extends TreeKind
+
+    case object TypeParameters extends TreeKind
+
+    /////////////// DECLARATIONS ///////////////
+    sealed trait Decl extends TreeKind {
+      override def debug_name: Option[String] = Some("Decl")
+    }
+
+    object Decl {
+      case object AssociatedTypeDef extends Decl
+
+      case object AssociatedTypeSig extends Decl
+
+      case object Class extends Decl
+
+      case object Def extends Decl
+
+      case object Effect extends Decl
+
+      case object Enum extends Decl
+
+      case object Instance extends Decl
+
+      case object Law extends Decl
+
+      case object Module extends Decl
+
+      case object Operation extends Decl
+
+      case object RestrictableEnum extends Decl
+
+      case object Signature extends Decl
+
+      case object TypeAlias extends Decl
+    }
+
+    /////////////// EXPRESSIONS  ///////////////
+    sealed trait Expr extends TreeKind {
+      override def debug_name: Option[String] = Some("Expr")
+    }
+
+    object Expr {
+
+      /**
+       * A marker kind used to wrap nested expressions.
+       * For instance on a binary expression "1 + 2" you would do
+       * Expr
+       *   Binary
+       *     Expr
+       *       LiteralNumber
+       *     Operator
+       *     Expr
+       *        LiteralNumber
+       */
+      case object Expr extends Expr
+
+      case object Ascribe extends Expr
+
+      case object Binary extends Expr
+
+      case object Block extends Expr
+
+      case object Call extends Expr
+
+      case object CheckedEffectCast extends Expr
+
+      case object CheckedTypeCast extends Expr
+
+      case object Do extends Expr
+
+      case object FixpointConstraint extends Expr
+
+      case object FixpointConstraintSet extends Expr
+
+      case object FixpointFrom extends Expr
```

Comment:
I notice that for a few other constructs. I have no problem with it, I am just asking how it works.

Reply:
The reasoning is that it is convenient to have a node in the syntax tree for each `FixpointFrom` or similar, to be able to `pickAll(TreeKind.Expr.FixpointFrom, ...` in the weeder and do further transformations/validation from there.

I can see why it might be a little misleading to extend `Expr` directly. The catch is that `Expr` really is just "a group of tree kinds that relate to expressions" rather than "the group of tree kinds that make up valid expressions".

Reply:
No its fine. Expr is expressions and things related to expressions.

Reply:
We can leave it or have ExprFragments

Reply:
Or rename the fragments to FixpointFromFragment

Reply:
Thoughts?